hosts/manager.go

```
@@ -27,30 +51,210 @@ type (
+					log := m.log.With(zap.Stringer("hk", host))
+					address, err := m.resolveHost(ctx, host)
+					if err != nil {
+						log.Error("failed to resolve host", zap.Error(err))
+						continue
+					}
+					err = m.scanHost(ctx, host, address)
+					if err != nil {
+						log.Error("failed to scan host", zap.Error(err))
+					}
+				}
+			}
+			if atomic.AddInt32(&activeThreads, -1) == 0 {
+				m.log.Info("scan finished", zap.Duration("elapsed", time.Since(start)))
+				close(doneChan)
+			}
+		}()
+	}
+
+	return doneChan
+}
+
+func (m *HostManager) resolveHost(ctx context.Context, hk types.PublicKey) (string, error) {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	defer cancel()
+
+	addrs, err := m.store.HostAddresses(ctx, hk)
+	if err != nil {
+		return "", err
+	} else if len(addrs) == 0 {
+		return "", errHostNoAddress
+	}
+
+	dedupe := make(map[string]net.IPAddr)
+	for _, na := range addrs {
+		host, _, err := net.SplitHostPort(na.Address)
+		if err != nil {
+			return "", err
+		}
+		ips, err := m.resolver.LookupIPAddr(ctx, host)
+		if err != nil {
+			return "", err
+		}
+		for _, ip := range ips {
+			dedupe[ip.String()] = ip
+		}
+	}
+
+	var networks []string
+	for _, addr := range dedupe {
+		ipRange := ipv6FilterRange
+		if addr.IP.To4() != nil {
+			ipRange = ipv4FilterRange
+		}
+
+		_, ipnet, err := net.ParseCIDR(fmt.Sprintf("%s/%d", addr.IP.String(), ipRange))
+		if err != nil {
+			return "", fmt.Errorf("failed to parse cidr: %w", err)
+		}
+		networks = append(networks, ipnet.String())
+	}
+
+	err = m.store.UpdateHostResolvedCIDRs(ctx, hk, networks)
+	if err != nil {
+		return "", fmt.Errorf("failed to update host resolved CIDRs: %w", err)
+	}
+
+	for _, addr := range dedupe {
+		if isPrivateIP(addr.IP) {
+			return "", fmt.Errorf("ip %q, %w", addr.IP, errHostPrivateIP)
+		}
+	}
+
+	for _, addr := range addrs {
+		if addr.Protocol == rhp4.ProtocolTCPSiaMux {
+			return addr.Address, nil
+		}
+	}
+	return "", errHostNoAddress
```

I think `resolveHost` tries to do too many things right now. Imo the only thing it should do is resolve all of the announced `chain.NetAddress`es of a host which potentially contains hostnames and then return a `[]chain.NetAddress` slice that only contains IP addresses. Private IP addresses and addresses for protocols which we don't support can be silently dropped. We don't care about those but I wouldn't punish a host for them.
Even returning an empty slice should be fine. The caller can handle that.

When we scan a host we should also probe every address and protocol. A host that is reachable via Siamux but not via Webtransport is potentially useless for clients and the scanning code is the only place where we have a chance to figure out that the host isn't reachable.

---

+1